diffcp/cone_program.py

```diff
@@ -179,30 +213,48 @@ def derivative(dA, db, dc, **kwargs):
 
     def adjoint_derivative(dx, dy, ds, **kwargs):
         """Applies adjoint of derivative at (A, b, c) to perturbations dx, dy, ds
-
         Args:
             dx: NumPy array representing perturbation in `x`
             dy: NumPy array representing perturbation in `y`
             ds: NumPy array representing perturbation in `s`
-
         Returns:
             (`dA`, `db`, `dc`), the result of applying the adjoint to the
             perturbations; the sparsity pattern of `dA` matches that of `A`.
         """
         dw = -(x @ dx + y @ dy + s @ ds)
         dz = np.concatenate(
             [dx, D_proj_dual_cone.rmatvec(dy + ds) - ds, np.array([dw])])
+
         if np.allclose(dz, 0):
             r = np.zeros(dz.shape)
-        else:
+        elif mode == "dense":
+            r = _diffcp._solve_adjoint_derivative_dense(M, MT, dz)
+        elif mode == "sparse":
+            rho = kwargs.get("rho", 1e-6)
+            it_ref_iters = kwargs.get("it_ref_iters", 5)
+            MT_iref = sparse.bmat([
+                [-sparse.eye(N), MT],
+                [M, rho * sparse.eye(N)]
+            ])
+            solve = splinalg.factorized(MT_iref.tocsc())
+            rhs = M @ dz
+            r = np.zeros(N)
+            # iterative refinement
+            for k in range(it_ref_iters):
+                residual = rhs - M @ (MT @ r)
+                if np.linalg.norm(residual) <= 1e-10:
+                    break
+                r = r + solve(np.append(np.zeros(N), residual))[N:]
+        elif mode == "lsqr":
```

Comment:
`else` and delete the other branch (since `mode` has already been validated)?

Reply:
done